src/NexusMods.DataModel/FileOriginRegistry.cs

```diff
@@ -109,28 +157,39 @@ public async ValueTask<DownloadId> RegisterFolder(AbsolutePath path, AArchiveMet
                 }).ToList(),
             MetaData = metaData
         };
+
+        _dataStore.AllIds(EntityCategory.DownloadMetadata);
         analysis.EnsurePersisted(_dataStore);
         return analysis.DownloadId;
-
     }
 
-    /// <inheritdoc />
-    public async ValueTask<DownloadAnalysis> Get(DownloadId id)
+    /// <summary>
+    ///     Build a Hash Table of all currently known files.
+    ///     We do this to deduplicate files between downloads.
+    /// </summary>
+    private HashCollection MakeKnownHashes()
     {
-        return _dataStore.Get<DownloadAnalysis>(IId.From(EntityCategory.DownloadMetadata, id.Value))!;
-    }
+        // Build a Hash Table of all currently known files. We do this to deduplicate files between downloads.
+        // TODO: This may need some benchmarking, it's unknown how well this scales across huge loadouts.
+        var archiveHashes = new Dictionary<ulong, DownloadId>();
+        var fileHashes = new HashSet<ulong>();
+        foreach (var ent in _dataStore.GetAll<DownloadAnalysis>(EntityCategory.DownloadMetadata)!)
+        {
+            // Note: Mods added from raw folders have a hash of '0', this is fine, just means there's nothing to dedupe.
+            archiveHashes[ent.Hash.Value] = ent.DownloadId;
+            foreach (var content in ent.Contents)
+                fileHashes.Add(content.Hash.Value);
+        }
```

Comment:
As we discussed in standup not all `ArchivedFiles` have a `DownloadAnalysis` so getting this doesn't cover all the game files or files that get Ingested.

You need to get all the `ArchiveFiles` directly.
You might want to add a method to `IFileStore` to do it instead of going directly for the database, just to keep things more organized.

Reply:
Is this desirable?

This means we are doing two trips to the database, rather than one. 

If we want to return a previously downloaded archive to do Archive level dedupe we need to query `DownloadAnalysis` so we can get a DownloadId, so that one's unavoidable. 

While it is true, that `DownloadAnalysis` does not contain every single file that the App has ever touched, consider the context of the situation we have here.

Hear me out:

- Our goal (and the original issue this PR addresses) is to deduplicate mod files. 
- The FileOriginRegistry's `DownloadAnalysis` table should contain all of the mod files.
    - Including those installed manually via `Add Mod`.
    - In other words every single file that's tied to a mod. (Not ingested)

So the issue you raised is the fact that we are not deduping with the original game files. 

*The whole point of a mod*, is to modify the game. If you are shipping unmodified game files as part of mods then you're
- [Breaking our own (and other sites') ToS](https://help.nexusmods.com/article/28-file-submission-guidelines) 
- Probably doing something wrong.

I currently pull the file hashes from there because I have to pull `DownloadAnalysis` anyway and it just happens to contain the set of all mod files.

Yes, there are also ingested files, but those are a separate issue. There is no dedupe mechanism for ingest (yet), and how we will use our ingest in the future is not yet clearly defined. (For example, the user may choose to delete ingested files, put them inside an existing mod. etc.) For ingest we can handle the deduping at the time the user decides what to do with the files, we can revisit this once we get there.